test/Npgsql.Tests/TypeMapperTests.cs

```diff
@@ -58,6 +60,95 @@ public async Task String_to_citext()
+
+        await using var command = new NpgsqlCommand("SELECT @p = 'hello'::citext", connection);
+        var parameter = new NpgsqlParameter("p", DbType.String)
+        {
+            Value = "HeLLo"
+        };
+        command.Parameters.Add(parameter);
+
+        Assert.That(command.ExecuteScalar(), Is.True);
+        Assert.That(parameter.DbType, Is.EqualTo(DbType.String));
+        Assert.That(parameter.NpgsqlDbType, Is.EqualTo(NpgsqlDbType.Text));
+        Assert.That(parameter.DataTypeName, Is.EqualTo("text"));
+    }
+
+    [Test]
+    public async Task Guid_to_custom_type()
+    {
+        await using var adminConnection = await OpenConnectionAsync();
+        var type = await GetTempTypeName(adminConnection);
+
+        var dataSourceBuilder = CreateDataSourceBuilder();
+        dataSourceBuilder.AddTypeInfoResolverFactory(new GuidTextConverterFactory(type));
+        await using var dataSource = dataSourceBuilder.Build();
+        await using var connection = await dataSource.OpenConnectionAsync();
+
+        await connection.ExecuteNonQueryAsync($"CREATE TYPE {type}");
+        await connection.ExecuteNonQueryAsync($"""
+            -- Input: cstring -> Custom type
+            CREATE FUNCTION {type}_in(cstring)
+            RETURNS {type}
+            AS 'textin'
+            LANGUAGE internal IMMUTABLE STRICT;
+
+            -- Output: Custom type -> cstring
+            CREATE FUNCTION {type}_out({type})
+            RETURNS cstring
+            AS 'textout'
+            LANGUAGE internal IMMUTABLE STRICT;
+
+            -- 3️⃣ Create wrappers for binary I/O
+            CREATE FUNCTION {type}_recv(internal)
+            RETURNS {type}
+            AS 'textrecv'
+            LANGUAGE internal IMMUTABLE STRICT;
+
+            CREATE FUNCTION {type}_send({type})
+            RETURNS bytea
+            AS 'textsend'
+            LANGUAGE internal IMMUTABLE STRICT;
+        """);
+
+        await connection.ExecuteNonQueryAsync($"""
+            CREATE TYPE {type} (
+                internallength = variable,
+                input = {type}_in,
+                output = {type}_out,
+                receive = {type}_recv,
+                send = {type}_send,
+                alignment = int4
+            );
+            CREATE CAST ({type} AS text) WITH INOUT AS IMPLICIT;
+            """);
+        await connection.ReloadTypesAsync();
+
+        var guid = Guid.NewGuid();
+        await using var command = new NpgsqlCommand($"SELECT @p::text = '{guid}'", connection);
+        var parameter = new NpgsqlParameter("p", DbType.Guid)
+        {
+            Value = guid
+        };
+        command.Parameters.Add(parameter);
+
+        Assert.That(command.ExecuteScalar(), Is.True);
+        Assert.That(parameter.DbType, Is.EqualTo(DbType.Guid));
+        Assert.That(parameter.NpgsqlDbType, Is.EqualTo(NpgsqlDbType.Uuid));
+        Assert.That(parameter.DataTypeName, Is.EqualTo("uuid"));
```

Comment:
@NinoFloris here I think we want this:
```suggestion
        Assert.That(parameter.NpgsqlDbType, Is.EqualTo(NpgsqlDbType.Unknow));
        Assert.That(parameter.DataTypeName, Is.EqualTo(type));
```

But since when NpgsqlParameter.DbType is set, _npgsqlDbType is also set I couldn't get this result.

To achieve this DbType would have to take priority in the get of NpgsqlDbType and DataTypeName, or remove this behavior of when one is set the other is also set.

I haven't test this properly but I suspect both approaches would cause regressions. 
I will take a closer look at this tomorrow, but if you have any ideias, or see something wrong, let me know.

Reply:
Yes they have their own storage now so they can be decoupled.